<!-- EDA Header / UX-style -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: linear-gradient(90deg, #EE7F2D 0%, #FFB56B 100%);
  color: #ffffff;
  padding: 18px 22px;
  border-radius: 14px;
  box-shadow: 0 10px 24px rgba(0,0,0,0.18);
  margin: 10px 0 14px 0;
">
  <div style="
    display:inline-block;
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    background: rgba(255,255,255,0.16);
    padding: 6px 10px;
    border-radius: 999px;
    margin-bottom: 10px;
  ">
    Give Me Some Credit • Exploratory Data Analysis
  </div>

  <div style="
    font-size: 30px;
    font-weight: 800;
    line-height: 1.15;
    margin: 6px 0 6px 0;
  ">
    Understanding Risk Signals Before Modeling
  </div>

  <div style="
    font-size: 15px;
    line-height: 1.5;
    opacity: .95;
    max-width: 980px;
  ">
    We explore <b>cs-training.csv</b> to map the default landscape, validate data quality, and uncover patterns that will guide feature preparation and threshold decisions.
  </div>
</div>

<!-- Optional section  -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 14px 16px;
  border-radius: 12px;
  margin: 0 0 10px 0;
  box-shadow: 0 10px 24px rgba(0,0,0,0.18);
">
  <div style="
    font-weight: 800;
    margin-bottom: 8px;
    letter-spacing: .02em;
  ">
    What we’ll cover
  </div>

  <ul style="
    margin: 0 0 0 18px;
    line-height: 1.65;
  ">
    Initial project analysis
    <li>Target distribution (default rate) and class imbalance</li>
    <li>Missing values, outliers, and inconsistent entries</li>
    <li>Descriptive statistics and variable profiles</li>
    <li>Early feature–target relationships to guide cleaning &amp; transformations</li>
  </ul>
</div>

<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: linear-gradient(90deg, #5B4661 0%, #918395 100%);
  color: #ffffff;
  padding: 16px 20px;
  border-radius: 14px;
  box-shadow: 0 10px 24px rgba(0,0,0,0.18);
  margin: 14px 0 12px 0;
">
  <div style="
    display:inline-block;
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    background: rgba(255,255,255,0.16);
    padding: 6px 10px;
    border-radius: 999px;
    margin-bottom: 10px;
  ">
    IMPORTING
  </div>

  <div style="
    font-size: 22px;
    font-weight: 800;
    line-height: 1.2;
    margin: 0 0 6px 0;
  ">
    Importing Libraries and Dataframes
  </div>

  <div style="
    font-size: 13px;
    line-height: 1.5;
    opacity: .92;
    max-width: 980px;
  ">
    R Libraries, training and disctionary dataframes
  </div>
</div>

<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Importing Libraries ~R
  </div>
</div>

In [82]:
library(magrittr)


<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Importing Dataframes ~csv
  </div>
</div>

In [83]:
train_path <- "../database/cs-training.csv"
cs_train <- read.csv(train_path)

#preview
head(cs_train, 10)

X,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30.59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60.89DaysPastDueNotWorse,NumberOfDependents
1,1,0.7661266,45,2,8.029821e-01,9120,13,0,6,0,2
2,0,0.9571510,40,0,1.218762e-01,2600,4,0,0,0,1
3,0,0.6581801,38,1,8.511338e-02,3042,2,1,0,0,0
4,0,0.2338098,30,0,3.604968e-02,3300,5,0,0,0,0
5,0,0.9072394,49,1,2.492570e-02,63588,7,0,1,0,0
6,0,0.2131787,74,0,3.756070e-01,3500,3,0,1,0,1
7,0,0.3056825,57,0,5.710000e+03,NA,8,0,3,0,0
8,0,0.7544636,39,0,2.099400e-01,3500,8,0,0,0,0
9,0,0.1169506,27,0,4.600000e+01,NA,2,0,0,0,NA
10,0,0.1891691,57,0,6.062909e-01,23684,9,0,4,0,2


In [84]:
dict_path <- "../database/data_dict.csv"

dict_df <- read.csv(dict_path, sep = ";", stringsAsFactors = FALSE, check.names = FALSE)
head(dict_df, 10)

Variable Name,Description,Type
SeriousDlqin2yrs,Person experienced 90 days past due delinquency or worse,Y/N
RevolvingUtilizationOfUnsecuredLines,Total balance on credit cards and personal lines of credit except real estate and no installment debt like car loans divided by the sum of credit limits,percentage
age,Age of borrower in years,integer
NumberOfTime30-59DaysPastDueNotWorse,Number of times borrower has been 30-59 days past due but no worse in the last 2 years.,integer
DebtRatio,"Monthly debt payments, alimony,living costs divided by monthy gross income",percentage
MonthlyIncome,Monthly income,real
NumberOfOpenCreditLinesAndLoans,Number of Open loans (installment like car loan or mortgage) and Lines of credit (e.g. credit cards),integer
NumberOfTimes90DaysLate,Number of times borrower has been 90 days or more past due.,integer
NumberRealEstateLoansOrLines,Number of mortgage and real estate loans including home equity lines of credit,integer
NumberOfTime60-89DaysPastDueNotWorse,Number of times borrower has been 60-89 days past due but no worse in the last 2 years.,integer


<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: linear-gradient(90deg, #5B4661 0%, #918395 100%);
  color: #ffffff;
  padding: 16px 20px;
  border-radius: 14px;
  box-shadow: 0 10px 24px rgba(0,0,0,0.18);
  margin: 14px 0 12px 0;
">
  <div style="
    display:inline-block;
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    background: rgba(255,255,255,0.16);
    padding: 6px 10px;
    border-radius: 999px;
    margin-bottom: 10px;
  ">
    EDA
  </div>

  <div style="
    font-size: 22px;
    font-weight: 800;
    line-height: 1.2;
    margin: 0 0 6px 0;
  ">
    Exploratory Data Analysis
  </div>

  <div style="
    font-size: 13px;
    line-height: 1.5;
    opacity: .92;
    max-width: 980px;
  ">
     Initial Exploration of Data Consistency
  </div>
</div>

<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Types and Column Names
  </div>
</div>

In [85]:
# Renaming ID column
names(cs_train)[1] <- "id"

# Renaming map -> Consistency in names
rename_map <- c(
  "SeriousDlqin2yrs"                    = "default_2y",
  "RevolvingUtilizationOfUnsecuredLines"= "util_unsecured",
  "age"                                 = "age_years",
  "NumberOfTime30.59DaysPastDueNotWorse"= "dpd_30_59_cnt",
  "DebtRatio"                           = "debt_ratio",
  "MonthlyIncome"                       = "monthly_income",
  "NumberOfOpenCreditLinesAndLoans"     = "open_credit_cnt",
  "NumberOfTimes90DaysLate"             = "dpd_90p_cnt",
  "NumberRealEstateLoansOrLines"        = "real_estate_cnt",
  "NumberOfTime60.89DaysPastDueNotWorse"= "dpd_60_89_cnt",
  "NumberOfDependents"                  = "dependents_cnt"
)

# Apply renaming
names(cs_train) <- ifelse(names(cs_train) %in% names(rename_map),
                          rename_map[names(cs_train)],
                          names(cs_train))

# Checking final column names ----
names(cs_train)

[1] "id"              "default_2y"      "util_unsecured"  "age_years"      
 [5] "dpd_30_59_cnt"   "debt_ratio"      "monthly_income"  "open_credit_cnt"
 [9] "dpd_90p_cnt"     "real_estate_cnt" "dpd_60_89_cnt"   "dependents_cnt"

<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Types
  </div>
</div>

In [86]:
str(cs_train)

'data.frame':	150000 obs. of  12 variables:
 $ id             : int  1 2 3 4 5 6 7 8 9 10 ...
 $ default_2y     : int  1 0 0 0 0 0 0 0 0 0 ...
 $ util_unsecured : num  0.766 0.957 0.658 0.234 0.907 ...
 $ age_years      : int  45 40 38 30 49 74 57 39 27 57 ...
 $ dpd_30_59_cnt  : int  2 0 1 0 1 0 0 0 0 0 ...
 $ debt_ratio     : num  0.803 0.1219 0.0851 0.036 0.0249 ...
 $ monthly_income : int  9120 2600 3042 3300 63588 3500 NA 3500 NA 23684 ...
 $ open_credit_cnt: int  13 4 2 5 7 3 8 8 2 9 ...
 $ dpd_90p_cnt    : int  0 0 1 0 0 0 0 0 0 0 ...
 $ real_estate_cnt: int  6 0 0 0 1 1 3 0 0 4 ...
 $ dpd_60_89_cnt  : int  0 0 0 0 0 0 0 0 0 0 ...
 $ dependents_cnt : int  2 1 0 0 0 1 0 0 NA 2 ...


<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Dataset Overview - Basic Metrics
  </div>
</div>

In [87]:
# Dataset overview
print('Dataset size: 1.rows x 2.columns:')
dim(cs_train)

cat('\nQuick descriptive statistics: numerical and categorical:\n')
summary(cs_train)

[1] "Dataset size: 1.rows x 2.columns:"


[1] 150000     12


Quick descriptive statistics: numerical and categorical:


       id           default_2y      util_unsecured       age_years    
 Min.   :     1   Min.   :0.00000   Min.   :    0.00   Min.   :  0.0  
 1st Qu.: 37501   1st Qu.:0.00000   1st Qu.:    0.03   1st Qu.: 41.0  
 Median : 75001   Median :0.00000   Median :    0.15   Median : 52.0  
 Mean   : 75001   Mean   :0.06684   Mean   :    6.05   Mean   : 52.3  
 3rd Qu.:112500   3rd Qu.:0.00000   3rd Qu.:    0.56   3rd Qu.: 63.0  
 Max.   :150000   Max.   :1.00000   Max.   :50708.00   Max.   :109.0  
                                                                      
 dpd_30_59_cnt      debt_ratio       monthly_income    open_credit_cnt 
 Min.   : 0.000   Min.   :     0.0   Min.   :      0   Min.   : 0.000  
 1st Qu.: 0.000   1st Qu.:     0.2   1st Qu.:   3400   1st Qu.: 5.000  
 Median : 0.000   Median :     0.4   Median :   5400   Median : 8.000  
 Mean   : 0.421   Mean   :   353.0   Mean   :   6670   Mean   : 8.453  
 3rd Qu.: 0.000   3rd Qu.:     0.9   3rd Qu.:   8249   3rd Qu.:11.000  


<!-- Note / comment box (minimal) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: rgba(170,170,170,0.10);
  border: 1px solid rgba(170,170,170,0.22);
  color: rgba(255,255,255,0.82);
  padding: 10px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .10em;
    text-transform: uppercase;
    opacity: .75;
    font-weight: 800;
    margin-bottom: 6px;
  ">
    Note - Anomalies identified (Aftercare)
  </div>
  <div style="font-size: 13px; line-height: 1.55;">
    <!-- Write your comment here -->

    - util_unsecured: Usage rate (between 0 and 1)
      - Max value is 50708. It must be between 0 and 1

    - age_years
      - Minimum value is 0, which doesn't make sense for age.
      - Credit is usually granted to those over 18 years of age.
      - Extreme age values ​​such as 109 years.
    
    - debt_ratio: Debt/monthly income ratio.
      - The average value is X and the maximum value is 933*x. This suggests glaring outliers, missing data, or income = 0.
    
    - monthly_income
      - High rate of null values
      - What is the difference between null values ​​and zero income?

    - dpd_90p_cnt: How many times has a payment been 90 days or more overdue.
      - The number 98 indicates that in this case there were 98 loans that were more than 90 days overdue. Which is somewhat unrealistic.

    - dependents_cnt: Number of dependents
      - Many null cases. Does this mean 0 or missing data?
  </div>
</div>

<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Duplicated Rows
  </div>
</div>

In [88]:
# Checking duplicate rows
cat('\nCounting how many IDs are repeated:\n')
dup_id  <- sum(duplicated(cs_train$id))
dup_id

cat('\nCounting how many complete rows are repeated (all columns are the same):\n')
dup_row <- sum(duplicated(cs_train))
dup_row


Counting how many IDs are repeated:


[1] 0


Counting how many complete rows are repeated (all columns are the same):


[1] 0

In [89]:
# Duplicated rows ignoring the ID column
cat('\nCounting how many complete rows (ignoring ID) are repeated (all columns are the same):\n')

dup_mask_no_id <- duplicated(cs_train[, setdiff(names(cs_train), "id")])

dup_n <- sum(dup_mask_no_id)
dup_pct <- 100 * dup_n / nrow(cs_train)

cat(sprintf("Repeated rows (ignoring ID): %d (%.2f%% of dataset)\n", dup_n, dup_pct))

dup_rows_no_id <- cs_train[
  dup_mask_no_id | duplicated(cs_train[, setdiff(names(cs_train), "id")], fromLast = TRUE),
]

dup_rows_no_id_sorted <- dup_rows_no_id[order(-dup_rows_no_id$age_years), ]
head(dup_rows_no_id_sorted, 20)


Counting how many complete rows (ignoring ID) are repeated (all columns are the same):
Repeated rows (ignoring ID): 609 (0.41% of dataset)


,id,default_2y,util_unsecured,age_years,dpd_30_59_cnt,debt_ratio,monthly_income,open_credit_cnt,dpd_90p_cnt,real_estate_cnt,dpd_60_89_cnt,dependents_cnt
78708,78708,0,0.0000000,99,0,0,NA,2,0,0,0,0
106913,106913,0,0.0000000,99,0,0,NA,2,0,0,0,0
131303,131303,0,0.0000000,96,0,0,NA,2,0,0,0,NA
132958,132958,0,0.0000000,96,0,0,NA,2,0,0,0,NA
31273,31273,0,0.9999999,92,0,0,NA,2,0,0,0,0
126061,126061,0,0.9999999,92,0,0,NA,2,0,0,0,0
6573,6573,0,0.0000000,91,0,0,NA,3,0,0,0,0
18324,18324,0,0.0000000,91,0,0,NA,3,0,0,0,0
25306,25306,0,0.9999999,91,0,0,NA,0,0,0,0,0
27922,27922,0,0.0000000,91,0,0,NA,5,0,0,0,0


<!-- Note / comment box (minimal) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: rgba(170,170,170,0.10);
  border: 1px solid rgba(170,170,170,0.22);
  color: rgba(255,255,255,0.82);
  padding: 10px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .10em;
    text-transform: uppercase;
    opacity: .75;
    font-weight: 800;
    margin-bottom: 6px;
  ">
    Note - Duplicate identification
  </div>
  <div style="font-size: 13px; line-height: 1.55;">
    <!-- Write your comment here -->

    0.41% of the dataset contains duplicate rows (excluding the ID). This may indeed represent coincidences of age x income x dependents x debt rate, etc. However, considering that the probability of repeating this data combination is very low for the same local database, we will consider it as duplicate data input errors. The data may have been entered twice by the responsible party while maintaining different IDs. We also need to consider that in the future, for the application of the models, we will remove the ID column, and duplicates can introduce false biases to the application.
    
    Furthermore, we can consider the possibility of one of the duplicate rows leaking into the validation: if a random train/valid split is performed and the same duplicate row exists, one copy may appear in the training and another in the test → the metric becomes artificially better (it seems that the model generalizes, but it only "recognized" the same record).
      
  </div>
</div>

<!-- Subsection header (ultra-compact) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: #5B4661;
  border: 1px solid rgba(255,255,255,0.18);
  color: #ffffff;
  padding: 8px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .12em;
    text-transform: uppercase;
    opacity: .92;
    font-weight: 800;
  ">
    Removing Duplicated Rows
  </div>
</div>

In [90]:
# Removing duplicated rows - keeping 1 of the duplicated rows
cs_train_dedup <- cs_train[!duplicated(cs_train[, setdiff(names(cs_train), "id")]), ]

In [92]:
# Checking again the duplicated rows
cat('\nCounting how many complete rows (ignoring ID) are repeated (all columns are the same):\n')

dup_mask_no_id <- duplicated(cs_train_dedup[, setdiff(names(cs_train_dedup), "id")])

dup_n <- sum(dup_mask_no_id)
dup_pct <- 100 * dup_n / nrow(cs_train_dedup)

cat(sprintf("Repeated rows (ignoring ID): %d (%.2f%% of dataset)\n", dup_n, dup_pct))

dup_rows_no_id <- cs_train_dedup[
  dup_mask_no_id | duplicated(cs_train_dedup[, setdiff(names(cs_train_dedup), "id")], fromLast = TRUE),
]

dup_rows_no_id_sorted <- dup_rows_no_id[order(-dup_rows_no_id$age_years), ]
head(dup_rows_no_id_sorted, 20)


Counting how many complete rows (ignoring ID) are repeated (all columns are the same):


Repeated rows (ignoring ID): 0 (0.00% of dataset)


id,default_2y,util_unsecured,age_years,dpd_30_59_cnt,debt_ratio,monthly_income,open_credit_cnt,dpd_90p_cnt,real_estate_cnt,dpd_60_89_cnt,dependents_cnt


<!-- Note / comment box (minimal) -->
<div style="
  font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, Helvetica, Arial;
  background: rgba(170,170,170,0.10);
  border: 1px solid rgba(170,170,170,0.22);
  color: rgba(255,255,255,0.82);
  padding: 10px 12px;
  border-radius: 12px;
  margin: 10px 0 12px 0;
">
  <div style="
    font-size: 12px;
    letter-spacing: .10em;
    text-transform: uppercase;
    opacity: .75;
    font-weight: 800;
    margin-bottom: 6px;
  ">
    Note - Duplicate rows removal
  </div>
  <div style="font-size: 13px; line-height: 1.55;">
    <!-- Write your comment here -->
    As described in the note above, we removed 609 lines (0.41%) with identical features and target to avoid bias and leakage.
  </div>
</div>